pybpodapi/bpod/emulator/emulator.py

```diff
@@ -0,0 +1,668 @@
+
+    def add_manual_override_event(self, event):
+        hardware_channels = self._hardware.channels
+        if isinstance(event, int):
+            try:
+                hardware_channels.get_event_name(event)
+            except IndexError as e:
+                raise EmulatorError(str(e))
+            else:
+                event_code = event
+        else:
+            if event not in hardware_channels.event_names:
+                raise EmulatorError('unknown event name')
+            event_code = hardware_channels.event_names.index(event)
+        self._manual_override_events.put(event_code)
+
+    def log_state_machine_info(self):
+        self._assert_state_machine()
+        sma = self._state_machine
+        highest_used_global_counter = sma.global_counters.get_max_index_used()
+        highest_used_global_timer = sma.global_timers.get_max_index_used()
+        highest_used_global_condition = sma.conditions.get_max_index_used()
+        highest_used_global_counter = (
+            0
+            if highest_used_global_counter is None
+            else highest_used_global_counter + 1
+        )
+        highest_used_global_timer = (
+            0
+            if highest_used_global_timer is None
+            else highest_used_global_timer + 1
+        )
+        highest_used_global_condition = (
+            0
+            if highest_used_global_condition is None
+            else highest_used_global_condition + 1
+        )
+        # State timer matrix (state timer transitions)
+        state_timer_matrix = []
+        for i in range(sma.total_states_added):
+            state_timer_matrix += [
+                sma.total_states_added
+                if math.isnan(sma.state_timer_matrix[i])
+                else sma.state_timer_matrix[i]
+            ]
+        logger.debug("STATE TIMER MATRIX: %s", state_timer_matrix)
+        # Input matrix (event-triggered transitions)
+        input_matrix = []
+        for i in range(sma.total_states_added):
+            state_transitions = sma.input_matrix[i]
+            n_transitions = len(state_transitions)
+            input_matrix += [n_transitions]
+            for transition in state_transitions:
+                input_matrix += [transition[0]]
+                dest_state = transition[1]
+                input_matrix += [
+                    sma.total_states_added
+                    if math.isnan(dest_state)
+                    else dest_state
+                ]
+        logger.debug("INPUT MATRIX: %s", input_matrix)
+        # Output matrix (hardware states)
+        output_matrix = []
+        global_timer_trigger_event_pos =  \
+            self.hardware.channels.events_positions.globalTimerTrigger
+        for i in range(sma.total_states_added):
+            hw_state = sma.output_matrix[i]
+            hw_state = [
+                evt
+                for evt in hw_state
+                if evt[0] < global_timer_trigger_event_pos
```

Comment:
Personally, I find list comprehensions are more redable if they are a one-liner. Maybe they can be done as:
```pyton
                pos = global_timer_trigger_event_pos
                hw_state = [evt for evt in hw_state if evt[0] < pos]
```